flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorTest.java

```diff
@@ -2536,6 +2484,61 @@ public void testDynamicSlotAllocation() throws Exception {
         }
     }
 
+    @Test
+    public void testReleasingJobResources() throws Exception {
+        AllocationID[] slots =
+                range(0, 5).mapToObj(i -> new AllocationID()).toArray(AllocationID[]::new);
+        try (TaskExecutorTestingContext ctx = createTaskExecutorTestingContext(slots.length)) {
+            ctx.start();
+            ResourceManagerId rmId;
+            {
+                CompletableFuture<Tuple3<ResourceID, InstanceID, SlotReport>>
+                        initialSlotReportFuture = new CompletableFuture<>();
+                rmId = createAndRegisterResourceManager(initialSlotReportFuture);
+                initialSlotReportFuture.get();
+            }
+
+            TaskExecutorGateway gateway =
+                    ctx.taskExecutor.getSelfGateway(TaskExecutorGateway.class);
+
+            for (int i = 0; i < slots.length; i++) {
+                requestSlot(
+                        gateway,
+                        jobId,
+                        slots[i],
+                        buildSlotID(i),
+                        ResourceProfile.UNKNOWN,
+                        ctx.jobMasterGateway.getAddress(),
+                        rmId);
+            }
+            ExecutionAttemptID attemptID =
+                    submitNoOpInvokableTask(
+                            slots[0], new TestingJobMasterGatewayBuilder().build(), gateway);
+            assertNotNull(ctx.changelogStoragesManager.getChangelogStoragesByJobId(jobId));
+            assertNotNull(ctx.metricGroup.getJobMetricsGroup(jobId));
+
+            // cancel tasks before releasing the slots - so that TM will release job resources on
+            // the last slot release
+            gateway.cancelTask(attemptID, timeout).get();
+            while (ctx.taskSlotTable.getTasks(jobId).hasNext()) {
```

Comment:
this access isn't thread-safe; it's backed by a plain HashMap but the test and TM thread are accessing it.
You need to run this within `ctx.taskExecutor.getMainThreadExecutableForTesting()`.

Reply:
Good point.